// @tybys/wasm-util/dist/wasm-util.esm-bundler.js
const _WebAssembly = typeof WebAssembly !== 'undefined'
    ? WebAssembly
    : typeof WXWebAssembly !== 'undefined'
        ? WXWebAssembly
        : undefined;
if (!_WebAssembly) {
    throw new Error('WebAssembly is not supported in this environment');
}

/* eslint-disable spaced-comment */

function validateObject(value, name) {
    if (value === null || typeof value !== 'object') {
        throw new TypeError(`${name} must be an object. Received ${value === null ? 'null' : typeof value}`);
    }
}
function validateArray(value, name) {
    if (!Array.isArray(value)) {
        throw new TypeError(`${name} must be an array. Received ${value === null ? 'null' : typeof value}`);
    }
}
function validateBoolean(value, name) {
    if (typeof value !== 'boolean') {
        throw new TypeError(`${name} must be a boolean. Received ${value === null ? 'null' : typeof value}`);
    }
}
function validateString(value, name) {
    if (typeof value !== 'string') {
        throw new TypeError(`${name} must be a string. Received ${value === null ? 'null' : typeof value}`);
    }
}
function validateFunction(value, name) {
    if (typeof value !== 'function') {
        throw new TypeError(`${name} must be a function. Received ${value === null ? 'null' : typeof value}`);
    }
}
function validateUndefined(value, name) {
    if (value !== undefined) {
        throw new TypeError(`${name} must be undefined. Received ${value === null ? 'null' : typeof value}`);
    }
}
function isPromiseLike(obj) {
    return !!(obj && (typeof obj === 'object' || typeof obj === 'function') && typeof obj.then === 'function');
}
function wrapInstanceExports(exports, mapFn) {
    const newExports = Object.create(null);
    Object.keys(exports).forEach(name => {
        const exportValue = exports[name];
        Object.defineProperty(newExports, name, {
            enumerable: true,
            value: mapFn(exportValue, name)
        });
    });
    return newExports;
}
function sleepBreakIf(delay, breakIf) {
    const start = Date.now();
    const end = start + delay;
    let ret = false;
    while (Date.now() < end) {
        if (breakIf()) {
            ret = true;
            break;
        }
    }
    return ret;
}
function unsharedSlice(view, start, end) {
    return ((typeof SharedArrayBuffer === 'function' && view.buffer instanceof SharedArrayBuffer) || (Object.prototype.toString.call(view.buffer.constructor) === '[object SharedArrayBuffer]'))
        ? view.slice(start, end)
        : view.subarray(start, end);
}

const ignoreNames = [
    'asyncify_get_state',
    'asyncify_start_rewind',
    'asyncify_start_unwind',
    'asyncify_stop_rewind',
    'asyncify_stop_unwind'
];
function tryAllocate(instance, wasm64, size, mallocName) {
    if (typeof instance.exports[mallocName] !== 'function' || size <= 0) {
        return {
            wasm64,
            dataPtr: 16,
            start: wasm64 ? 32 : 24,
            end: 1024
        };
    }
    const malloc = instance.exports[mallocName];
    const dataPtr = wasm64 ? Number(malloc(BigInt(16) + BigInt(size))) : malloc(8 + size);
    if (dataPtr === 0) {
        throw new Error('Allocate asyncify data failed');
    }
    return wasm64
        ? { wasm64, dataPtr, start: dataPtr + 16, end: dataPtr + 16 + size }
        : { wasm64, dataPtr, start: dataPtr + 8, end: dataPtr + 8 + size };
}
/** @public */
class Asyncify {
    constructor() {
        this.value = undefined;
        this.exports = undefined;
        this.dataPtr = 0;
    }
    init(memory, instance, options) {
        var _a, _b;
        if (this.exports) {
            throw new Error('Asyncify has been initialized');
        }
        if (!(memory instanceof _WebAssembly.Memory)) {
            throw new TypeError('Require WebAssembly.Memory object');
        }
        const exports = instance.exports;
        for (let i = 0; i < ignoreNames.length; ++i) {
            if (typeof exports[ignoreNames[i]] !== 'function') {
                throw new TypeError('Invalid asyncify wasm');
            }
        }
        let address;
        const wasm64 = Boolean(options.wasm64);
        if (!options.tryAllocate) {
            address = {
                wasm64,
                dataPtr: 16,
                start: wasm64 ? 32 : 24,
                end: 1024
            };
        }
        else {
            if (options.tryAllocate === true) {
                address = tryAllocate(instance, wasm64, 4096, 'malloc');
            }
            else {
                address = tryAllocate(instance, wasm64, (_a = options.tryAllocate.size) !== null && _a !== void 0 ? _a : 4096, (_b = options.tryAllocate.name) !== null && _b !== void 0 ? _b : 'malloc');
            }
        }
        this.dataPtr = address.dataPtr;
        if (wasm64) {
            new BigInt64Array(memory.buffer, this.dataPtr).set([BigInt(address.start), BigInt(address.end)]);
        }
        else {
            new Int32Array(memory.buffer, this.dataPtr).set([address.start, address.end]);
        }
        this.exports = this.wrapExports(exports, options.wrapExports);
        const asyncifiedInstance = Object.create(_WebAssembly.Instance.prototype);
        Object.defineProperty(asyncifiedInstance, 'exports', { value: this.exports });
        // Object.setPrototypeOf(instance, Instance.prototype)
        return asyncifiedInstance;
    }
    assertState() {
        if (this.exports.asyncify_get_state() !== 0 /* AsyncifyState.NONE */) {
            throw new Error('Asyncify state error');
        }
    }
    wrapImportFunction(f) {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const _this = this;
        return (function () {
            // eslint-disable-next-line no-unreachable-loop
            while (_this.exports.asyncify_get_state() === 2 /* AsyncifyState.REWINDING */) {
                _this.exports.asyncify_stop_rewind();
                return _this.value;
            }
            _this.assertState();
            const v = f.apply(this, arguments);
            if (!isPromiseLike(v))
                return v;
            _this.exports.asyncify_start_unwind(_this.dataPtr);
            _this.value = v;
        });
    }
    wrapImports(imports) {
        const importObject = {};
        Object.keys(imports).forEach(k => {
            const mod = imports[k];
            const newModule = {};
            Object.keys(mod).forEach(name => {
                const importValue = mod[name];
                if (typeof importValue === 'function') {
                    newModule[name] = this.wrapImportFunction(importValue);
                }
                else {
                    newModule[name] = importValue;
                }
            });
            importObject[k] = newModule;
        });
        return importObject;
    }
    wrapExportFunction(f) {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const _this = this;
        return (async function () {
            _this.assertState();
            let ret = f.apply(this, arguments);
            while (_this.exports.asyncify_get_state() === 1 /* AsyncifyState.UNWINDING */) {
                _this.exports.asyncify_stop_unwind();
                _this.value = await _this.value;
                _this.assertState();
                _this.exports.asyncify_start_rewind(_this.dataPtr);
                ret = f.call(this);
            }
            _this.assertState();
            return ret;
        });
    }
    wrapExports(exports, needWrap) {
        return wrapInstanceExports(exports, (exportValue, name) => {
            let ignore = ignoreNames.indexOf(name) !== -1 || typeof exportValue !== 'function';
            if (Array.isArray(needWrap)) {
                ignore = ignore || (needWrap.indexOf(name) === -1);
            }
            return ignore ? exportValue : this.wrapExportFunction(exportValue);
        });
    }
}

function validateImports(imports) {
    if (imports && typeof imports !== 'object') {
        throw new TypeError('imports must be an object or undefined');
    }
}
function fetchWasm(urlOrBuffer, imports) {
    if (typeof wx !== 'undefined' && typeof __wxConfig !== 'undefined') {
        return _WebAssembly.instantiate(urlOrBuffer, imports);
    }
    return fetch(urlOrBuffer)
        .then(response => response.arrayBuffer())
        .then(buffer => _WebAssembly.instantiate(buffer, imports));
}
/** @public */
function load(wasmInput, imports) {
    validateImports(imports);
    imports = imports !== null && imports !== void 0 ? imports : {};
    let source;
    if (wasmInput instanceof ArrayBuffer || ArrayBuffer.isView(wasmInput)) {
        return _WebAssembly.instantiate(wasmInput, imports);
    }
    if (wasmInput instanceof _WebAssembly.Module) {
        return _WebAssembly.instantiate(wasmInput, imports).then((instance) => {
            return { instance, module: wasmInput };
        });
    }
    if (typeof wasmInput !== 'string' && !(wasmInput instanceof URL)) {
        throw new TypeError('Invalid source');
    }
    if (typeof _WebAssembly.instantiateStreaming === 'function') {
        let responsePromise;
        try {
            responsePromise = fetch(wasmInput);
            source = _WebAssembly.instantiateStreaming(responsePromise, imports).catch(() => {
                return fetchWasm(wasmInput, imports);
            });
        }
        catch (_) {
            source = fetchWasm(wasmInput, imports);
        }
    }
    else {
        source = fetchWasm(wasmInput, imports);
    }
    return source;
}
/** @public */
function asyncifyLoad(asyncify, urlOrBuffer, imports) {
    validateImports(imports);
    imports = imports !== null && imports !== void 0 ? imports : {};
    const asyncifyHelper = new Asyncify();
    imports = asyncifyHelper.wrapImports(imports);
    return load(urlOrBuffer, imports).then(source => {
        var _a;
        const memory = source.instance.exports.memory || ((_a = imports.env) === null || _a === void 0 ? void 0 : _a.memory);
        return { module: source.module, instance: asyncifyHelper.init(memory, source.instance, asyncify) };
    });
}
/** @public */
function loadSync(wasmInput, imports) {
    validateImports(imports);
    imports = imports !== null && imports !== void 0 ? imports : {};
    let module;
    if ((wasmInput instanceof ArrayBuffer) || ArrayBuffer.isView(wasmInput)) {
        module = new _WebAssembly.Module(wasmInput);
    }
    else if (wasmInput instanceof WebAssembly.Module) {
        module = wasmInput;
    }
    else {
        throw new TypeError('Invalid source');
    }
    const instance = new _WebAssembly.Instance(module, imports);
    const source = { instance, module };
    return source;
}
/** @public */
function asyncifyLoadSync(asyncify, buffer, imports) {
    var _a;
    validateImports(imports);
    imports = imports !== null && imports !== void 0 ? imports : {};
    const asyncifyHelper = new Asyncify();
    imports = asyncifyHelper.wrapImports(imports);
    const source = loadSync(buffer, imports);
    const memory = source.instance.exports.memory || ((_a = imports.env) === null || _a === void 0 ? void 0 : _a.memory);
    return { module: source.module, instance: asyncifyHelper.init(memory, source.instance, asyncify) };
}

const CHAR_DOT = 46; /* . */
const CHAR_FORWARD_SLASH = 47; /* / */
const CHAR_BACKWARD_SLASH = 92; /* \ */
const CHAR_COLON = 58; /* : */
const CHAR_UPPERCASE_A = 65; /* A */
const CHAR_UPPERCASE_Z = 90; /* Z */
const CHAR_LOWERCASE_A = 97; /* a */
const CHAR_LOWERCASE_Z = 122; /* z */
function isPathSeparatorWin(code) {
    return code === CHAR_FORWARD_SLASH || code === CHAR_BACKWARD_SLASH;
}
function isWindowsDeviceRoot(code) {
    return (code >= CHAR_UPPERCASE_A && code <= CHAR_UPPERCASE_Z) ||
        (code >= CHAR_LOWERCASE_A && code <= CHAR_LOWERCASE_Z);
}
const _isWin32 = typeof process !== 'undefined' && process.platform === 'win32';
/**
 * Windows variant of `resolve()`. Mirrors Node's `path.win32.resolve`
 * semantics enough for the WASI shim's needs (drive-letter absolutes
 * + UNC paths + per-drive cwd lookups via `process.env`/`process.cwd`).
 *
 * Why this is needed: on Windows, `FileDescriptor.realPath` is the
 * host realpath returned by `fs.realpathSync(realPath, 'utf8')` — the
 * backslash form `D:\…`. The POSIX-only `resolveImpl` below only
 * treats `/` as a separator, reads `D` as non-`/`, decides realPath
 * is "relative", and produces a garbage joined path. Downstream
 * `fs.openSync(garbage)` then returns `EINVAL` and the WASI caller
 * sees a permanent failure. This function gives us the correct
 * Windows-style resolution without taking a Node-only `path` import
 * (which would break browser bundles of this package).
 *
 * Cribbed from Node's `lib/path.js` `win32.resolve()` (MIT-licensed),
 * trimmed to what WASI realpath resolution actually exercises.
 */
function resolveWin32(args) {
    let resolvedDevice = '';
    let resolvedTail = '';
    let resolvedAbsolute = false;
    for (let i = args.length - 1; i >= -1; i--) {
        let path;
        if (i >= 0) {
            path = args[i];
            validateString(path, 'path');
            if (path.length === 0)
                continue;
        }
        else if (resolvedDevice.length === 0) {
            path = (typeof process !== 'undefined' && typeof process.cwd === 'function')
                ? process.cwd()
                : '';
        }
        else {
            // Look up per-drive cwd via the `=X:` env var convention; fall back
            // to the global cwd if absent.
            const envKey = `=${resolvedDevice}`;
            const env = (typeof process !== 'undefined') ? process.env : undefined;
            path = (env && typeof env[envKey] === 'string')
                ? env[envKey]
                : (typeof process !== 'undefined' && typeof process.cwd === 'function')
                    ? process.cwd()
                    : '';
            if (path === undefined ||
                (path.slice(0, 2).toLowerCase() !== resolvedDevice.toLowerCase() &&
                    path.charCodeAt(2) === CHAR_BACKWARD_SLASH)) {
                path = `${resolvedDevice}\\`;
            }
        }
        const len = path.length;
        let rootEnd = 0;
        let device = '';
        let isAbsolute = false;
        const code = path.charCodeAt(0);
        if (len === 1) {
            if (isPathSeparatorWin(code)) {
                rootEnd = 1;
                isAbsolute = true;
            }
        }
        else if (isPathSeparatorWin(code)) {
            isAbsolute = true;
            if (isPathSeparatorWin(path.charCodeAt(1))) {
                // UNC path: `\\server\share\…`
                let j = 2;
                let last = j;
                while (j < len && !isPathSeparatorWin(path.charCodeAt(j)))
                    j++;
                if (j < len && j !== last) {
                    const firstPart = path.slice(last, j);
                    last = j;
                    while (j < len && isPathSeparatorWin(path.charCodeAt(j)))
                        j++;
                    if (j < len && j !== last) {
                        last = j;
                        while (j < len && !isPathSeparatorWin(path.charCodeAt(j)))
                            j++;
                        if (j === len || j !== last) {
                            device = `\\\\${firstPart}\\${path.slice(last, j)}`;
                            rootEnd = j;
                        }
                    }
                }
            }
            else {
                rootEnd = 1;
            }
        }
        else if (isWindowsDeviceRoot(code) && path.charCodeAt(1) === CHAR_COLON) {
            device = path.slice(0, 2);
            rootEnd = 2;
            if (len > 2 && isPathSeparatorWin(path.charCodeAt(2))) {
                isAbsolute = true;
                rootEnd = 3;
            }
        }
        if (device.length > 0) {
            if (resolvedDevice.length > 0) {
                if (device.toLowerCase() !== resolvedDevice.toLowerCase())
                    continue;
            }
            else {
                resolvedDevice = device;
            }
        }
        if (resolvedAbsolute) {
            if (resolvedDevice.length > 0)
                break;
        }
        else {
            resolvedTail = `${path.slice(rootEnd)}\\${resolvedTail}`;
            resolvedAbsolute = isAbsolute;
            if (isAbsolute && resolvedDevice.length > 0)
                break;
        }
    }
    resolvedTail = normalizeString(resolvedTail, !resolvedAbsolute, '\\', isPathSeparatorWin);
    return resolvedDevice + (resolvedAbsolute ? '\\' : '') + resolvedTail || '.';
}
function isPosixPathSeparator(code) {
    return code === CHAR_FORWARD_SLASH;
}
function normalizeString(path, allowAboveRoot, separator, isPathSeparator) {
    let res = '';
    let lastSegmentLength = 0;
    let lastSlash = -1;
    let dots = 0;
    let code = 0;
    for (let i = 0; i <= path.length; ++i) {
        if (i < path.length) {
            code = path.charCodeAt(i);
        }
        else if (isPathSeparator(code)) {
            break;
        }
        else {
            code = CHAR_FORWARD_SLASH;
        }
        if (isPathSeparator(code)) {
            if (lastSlash === i - 1 || dots === 1) ;
            else if (dots === 2) {
                if (res.length < 2 || lastSegmentLength !== 2 ||
                    res.charCodeAt(res.length - 1) !== CHAR_DOT ||
                    res.charCodeAt(res.length - 2) !== CHAR_DOT) {
                    if (res.length > 2) {
                        const lastSlashIndex = res.indexOf(separator);
                        if (lastSlashIndex === -1) {
                            res = '';
                            lastSegmentLength = 0;
                        }
                        else {
                            res = res.slice(0, lastSlashIndex);
                            lastSegmentLength =
                                res.length - 1 - res.indexOf(separator);
                        }
                        lastSlash = i;
                        dots = 0;
                        continue;
                    }
                    else if (res.length !== 0) {
                        res = '';
                        lastSegmentLength = 0;
                        lastSlash = i;
                        dots = 0;
                        continue;
                    }
                }
                if (allowAboveRoot) {
                    res += res.length > 0 ? `${separator}..` : '..';
                    lastSegmentLength = 2;
                }
            }
            else {
                if (res.length > 0) {
                    res += `${separator}${path.slice(lastSlash + 1, i)}`;
                }
                else {
                    res = path.slice(lastSlash + 1, i);
                }
                lastSegmentLength = i - lastSlash - 1;
            }
            lastSlash = i;
            dots = 0;
        }
        else if (code === CHAR_DOT && dots !== -1) {
            ++dots;
        }
        else {
            dots = -1;
        }
    }
    return res;
